skrifa/src/meta/charmap.rs

```diff
@@ -0,0 +1,415 @@
+//! simple text containing diacritics and ligatures.
+//!
+//! Comprehensive mapping of characters to positioned glyphs requires a process called
+//! shaping. For more detail, see: [Why do I need a shaping engine?](https://harfbuzz.github.io/why-do-i-need-a-shaping-engine.html)
+
+use read_fonts::{
+    tables::cmap::{self, Cmap, Cmap12, Cmap14, Cmap4, CmapSubtable, EncodingRecord, PlatformId},
+    types::GlyphId,
+    FontData, TableProvider,
+};
+
+pub use read_fonts::tables::cmap::MapVariant;
+
+/// Mapping of characters to nominal glyph identifiers.
+///
+/// The mappings are derived from the [cmap](https://learn.microsoft.com/en-us/typography/opentype/spec/cmap)
+/// table.
+///
+/// ## Selection strategy
+///
+/// Fonts may contain multiple subtables in various formats supporting different encodings. The selection
+/// strategy implemented here is designed to choose mappings that capture the broadest available Unicode
+/// coverage:
+///
+/// * Unicode characters: a symbol mapping subtable is selected if available. Otherwise, subtables supporting
+/// the Unicode full repertoire or Basic Multilingual Plane (BMP) are preferred, in that order. Formats
+/// [4](https://learn.microsoft.com/en-us/typography/opentype/spec/cmap#format-4-segment-mapping-to-delta-values)
+/// and [12](https://learn.microsoft.com/en-us/typography/opentype/spec/cmap#format-12-segmented-coverage) are
+/// supported.
+///
+/// * Unicode variation sequences: these are provided by a format
+/// [14](https://learn.microsoft.com/en-us/typography/opentype/spec/cmap#format-14-unicode-variation-sequences)
+/// subtable.
+///
+#[derive(Clone, Default)]
+pub struct Charmap<'a> {
+    map: Option<Map<'a>>,
+    variant_map: Option<Cmap14<'a>>,
+}
+
+impl<'a> Charmap<'a> {
+    /// Creates a new character map from the given font.
+    pub fn new(font: &impl TableProvider<'a>) -> Self {
+        if let Ok(cmap) = font.cmap() {
+            let selection = MappingSelection::new(&cmap);
+            Self {
+                map: selection.map.map(|subtable| Map {
+                    subtable,
+                    is_symbol: selection.mapping_index.map_is_symbol,
+                }),
+                variant_map: selection.variant_map,
+            }
+        } else {
+            Self::default()
+        }
+    }
+
+    /// Returns true if a suitable Unicode character mapping is available.
+    pub fn has_map(&self) -> bool {
+        self.map.is_some()
+    }
+
+    /// Returns true if a symbol mapping was selected.
+    pub fn is_symbol(&self) -> bool {
+        self.map.as_ref().map(|x| x.is_symbol).unwrap_or(false)
+    }
+
+    /// Returns true if a Unicode variation sequence mapping is available.
+    pub fn has_variant_map(&self) -> bool {
+        self.variant_map.is_some()
+    }
+
+    /// Maps a character to a nominal glyph identifier. Returns `None` if a mapping does
+    /// not exist.
+    pub fn map(&self, ch: impl Into<u32>) -> Option<GlyphId> {
```

Comment:
More references:
* https://bugs.chromium.org/p/chromium/issues/detail?id=589340 discussion on differences when moving to HarfBuzz glyph lookup functions, and at the time found that Firefox does not seem to support symbol cmaps.
* But also: https://github.com/harfbuzz/harfbuzz/issues/1918#issuecomment-693630196 - where @behdad explains that CoreText does not prioritise symbol.

I had two things in mind: One would be: 
1) We leave it up the application control this and provide a `map_symbol()` method in addition where mapping through the symbol table is performed when one is found, and always select a non-symbol cmap for `fn map()`. 
2) Or we leave as is with the reasoning that if someone puts a symbol table in their font, they probably want it to work for looking up symbols primarily - if there are other glyphs in that font, it could be a regular cmap lookup with PUA codepoint lookup results.

Option 2. has less magic, but might not be 100% required at the moment, given that we did not encounter any issues with that in Chrome to my knowledge for the past couple years.

Reply:
Thank you for the additional context! Shipping for several years in both Chrome and HarfBuzz feels like really strong precedent so my preference would be to go with option 2 and not diverge from current behavior unless we find some reason to do otherwise.

Reply:
Yes, ok with me. I can't foresee that precisely yet, but due to requirements of another client, we may need more control, similar to `FT_Select_Charmap` here in the future, but we can revisit this then.

Reply:
Happy to revisit if a client requests more fine grained control.